src/v/config/property.h

```diff
@@ -956,5 +965,157 @@ class hidden_when_default_property : public property<T> {
+     *   - A = [z, x, y] is restricted
+     *   - A = [y, z] is NOT restricted
+     *
+     */
+    using val_t = typename value_type<T>::type;
+    using val_container_t = std::vector<val_t>;
+    using restrict_check_t = std::function<bool(const val_t&)>;
+    using restrict_variant_t
+      = std::variant<val_t, val_container_t, restrict_check_t>;
+
+public:
+    /**
+     * Construct an enterprise property.
+     *
+     * Largely a pass-through to the wrapped property's constructor, with the
+     * addition of a variant parameter describing restricted values, which may
+     * take the type
+     *   - val_t: that value is restricted
+     *   - vector<val_t>: each element is restricted
+     *   - std::function<bool(val_t)>: predicate for whether some val_t is
+     *     restricted
+     *
+     * NOTE: Value restrictions must never conflict with a config's default
+     * value. This condition is enforced by an assertion in the constructor.
+     *
+     */
+    template<typename... Args>
+    enterprise(
+      config_store& conf, restrict_variant_t restricted, Args&&... args)
+      : P(conf, std::forward<Args>(args)...)
+      , _restriction(std::move(restricted)) {
+        assert_no_default_conflict();
+    }
+
+    /**
+     * Decodes the given YAML node into the underlying property's value_type and
+     * checks whether that value should be restricted to enterprise clusters
+     * based on the unwrapped value restrictions described at construction.
+     */
+    std::optional<validation_error> check_restricted(YAML::Node n) const final {
+        auto v = std::move(n.as<T>());
+        if (check_restricted(v)) {
+            return std::make_optional<validation_error>(
+              P::name().data(),
+              ssx::sformat(
+                "'{}' is restricted to enterprise licensed clusters", v));
+        }
+        return std::nullopt;
+    }
+
+private:
+    void assert_no_default_conflict() const {
+        vassert(
+          !check_restricted(this->default_value()),
+          "Enterprise properties must not restrict the default value of the "
+          "underlying property!");
+    }
+
+    bool check_restricted(const T& setting) const {
+        // depending on how the restriction was defined, construct an applicable
+        // check function for bare instances of the underlying value type
+        auto restriction_check = [this](const val_t& v) -> bool {
+            return ss::visit(
+              _restriction,
+              [&v](const val_t& val) { return v == val; },
+              [&v](const val_container_t& vals) {
+                  return std::ranges::find(vals, v) != vals.end();
+              },
+              [&v](const restrict_check_t& check) { return check(v); });
+        };
+
+        if constexpr (detail::Array<T>) {
+            return std::ranges::any_of(
+              setting, [&restriction_check](const auto& v) {
+                  return restriction_check(v);
+              });
```

Comment:
nitpick:
```suggestion
            return std::ranges::any_of(setting, restriction_check);
```

Reply:
weird, i recall changing this (and the other `ranges::end` one) on my branch. i wonder whether i stashed them during rebase and forgot to reapply... anyway good for quick followup